packages/sdk/src/transactions/Inscriber.ts

```diff
@@ -0,0 +1,300 @@
+  network: Network
+
+  mediaType: string
+  mediaContent: string
+  meta?: NestedObject
+  postage: number
+
+  address: string
+  publicKey: string
+  destinationAddress: string
+  changeAddress: string
+  inputsToSign: InputsToSign
+
+  private ready = false
+
+  private xKey!: string
+  private commitAddress: string | null = null
+  private payment: bitcoin.payments.Payment | null = null
+  private suitableUnspent: UTXOLimited | null = null
+  private recovery = false
+  private safeMode: OnOffUnion
+  private encodeMetadata: boolean
+  private previewMode = false
+
+  private witnessScripts: Record<"inscription" | "recovery", Buffer | null> = {
+    inscription: null,
+    recovery: null
+  }
+  private taprootTree!: [Tapleaf, Tapleaf]
+
+  constructor({
+    network,
+    feeRate,
+    postage,
+    mediaContent,
+    mediaType,
+    publicKey,
+    outputs = [],
+    encodeMetadata = false,
+    changeAddress,
+    destination,
+    safeMode,
+    meta
+  }: InscriberArgOptions) {
+    const { xkey, address } = getAddressesFromPublicKey(publicKey, network, "p2tr")[0]
+    super({
+      address: address!,
+      feeRate,
+      network,
+      publicKey,
+      outputs,
+      inscriberMode: true
+    })
+    if (!publicKey || !changeAddress || !destination || !mediaContent) {
+      throw new Error("Invalid options provided")
+    }
+
+    this.publicKey = publicKey
+    this.feeRate = feeRate
+    this.mediaType = mediaType
+    this.network = network
+    this.changeAddress = changeAddress
+    this.destinationAddress = destination
+    this.mediaContent = mediaContent
+    this.meta = meta
+    this.postage = postage
+    this.outputs = outputs
+    this.safeMode = !safeMode ? "on" : safeMode
+    this.encodeMetadata = encodeMetadata
+
+    if (!xkey) {
+      throw new Error("Failed to derive xKey from the provided public key")
+    }
+
+    this.xKey = xkey
+    this.address = address!
+    this.inputsToSign = {
+      address: this.address,
+      signingIndexes: []
+    }
+  }
+
+  private getMetadata() {
+    return this.meta && this.encodeMetadata ? encodeObject(this.meta) : this.meta
+  }
+
+  async build() {
+    if (!this.suitableUnspent || !this.payment) {
+      throw new Error("Failed to build PSBT. Transaction not ready")
+    }
+
+    this.inputs = [
+      {
+        type: "taproot",
+        hash: this.suitableUnspent.txid,
+        index: this.suitableUnspent.n,
+        tapInternalKey: Buffer.from(this.xKey, "hex"),
+        witnessUtxo: {
+          script: this.payment.output!,
+          value: this.suitableUnspent.sats
+        },
+        tapLeafScript: [
+          {
+            leafVersion: this.payment.redeemVersion!,
+            script: this.payment.redeem!.output!,
+            controlBlock: this.payment.witness![this.payment.witness!.length - 1]
+          }
+        ]
+      }
+    ]
+
+    this.inputsToSign.signingIndexes.push(0) // hardcoding because there will always be one input
```

Comment:
maybe a safer way is just to use `this.inputs.length - 1`? incase `this.inputs` changes for some reason, it will always be correct.